File: src/main/java/io/github/hdzitao/editstarters/buildsystem/gradle/GradleBuildSystem.java
package io.github.hdzitao.editstarters.buildsystem.gradle;

import com.intellij.openapi.actionSystem.CommonDataKeys;
import com.intellij.openapi.actionSystem.DataContext;
import com.intellij.openapi.externalSystem.model.DataNode;
import com.intellij.openapi.externalSystem.model.ProjectKeys;
import com.intellij.openapi.externalSystem.model.project.LibraryData;
import com.intellij.openapi.externalSystem.model.project.ProjectData;
import com.intellij.openapi.externalSystem.util.ExternalSystemApiUtil;
import com.intellij.openapi.project.Project;
import com.intellij.psi.PsiFile;
import io.github.hdzitao.editstarters.buildsystem.BuildSystem;
import io.github.hdzitao.editstarters.dependency.Dependency;
import org.jetbrains.kotlin.psi.KtFile;
import org.jetbrains.plugins.gradle.util.GradleConstants;
import org.jetbrains.plugins.groovy.lang.psi.GroovyFile;

import java.util.List;
import java.util.Objects;
import java.util.stream.Collectors;

/**
 * gradle构建系统
 *
 * @version 3.2.0
 */
public class GradleBuildSystem extends BuildSystem {

    private GradleBuildSystem(AbstractBuildGradle<?> buildGradle, List<Dependency> dependencies) {
        super(buildGradle, dependencies);
    }

    /**
     * 根据文件名构建gradle build system
     */
    public static GradleBuildSystem from(DataContext context) {
        PsiFile psiFile = Objects.requireNonNull(context.getData(CommonDataKeys.PSI_FILE));
        Project project = Objects.requireNonNull(context.getData(CommonDataKeys.PROJECT));

        String name = psiFile.getName();
        AbstractBuildGradle<?> buildGradle = switch (name) {
            case GradleConstants.DEFAULT_SCRIPT_NAME -> new BuildGradle(project, (GroovyFile) psiFile);
            case GradleConstants.KOTLIN_DSL_SCRIPT_NAME -> new BuildGradleKts(project, (KtFile) psiFile);
            default -> throw new UnsupportedOperationException("Not support extension!");
        };

        DataNode<ProjectData> projectData = Objects.requireNonNull(ExternalSystemApiUtil.findProjectNode(
                project,
                GradleConstants.SYSTEM_ID,
                Objects.requireNonNull(project.getBasePath())));
        List<Dependency> dependencies = projectData.getChildren().stream()
                .filter(node -> ProjectKeys.LIBRARY.equals(node.getKey()))
                .map(node -> (LibraryData) node.getData())
                .map(lib -> new Dependency(lib.getGroupId(), lib.getArtifactId(), lib.getVersion()))
                .collect(Collectors.toList());

        return new GradleBuildSystem(buildGradle, dependencies);
    }
}

File: src/main/kotlin/io/github/hdzitao/editstarters/buildsystem/gradle/GradleBuildSystem.kt
package io.github.hdzitao.editstarters.buildsystem.gradle

import com.intellij.openapi.actionSystem.CommonDataKeys
import com.intellij.openapi.actionSystem.DataContext
import com.intellij.openapi.externalSystem.model.DataNode
import com.intellij.openapi.externalSystem.model.ProjectKeys
import com.intellij.openapi.externalSystem.model.project.LibraryData
import com.intellij.openapi.externalSystem.util.ExternalSystemApiUtil
import io.github.hdzitao.editstarters.buildsystem.BuildSystem
import io.github.hdzitao.editstarters.dependency.Dependency
import io.github.hdzitao.editstarters.ui.ShowErrorException
import lombok.SneakyThrows
import org.apache.commons.lang3.StringUtils
import org.jetbrains.kotlin.psi.KtFile
import org.jetbrains.plugins.gradle.util.GradleConstants
import org.jetbrains.plugins.groovy.lang.psi.GroovyFile
import java.util.stream.Collectors

/**
 * gradle构建系统
 *
 * @version 3.2.0
 */
class GradleBuildSystem private constructor(buildGradle: AbstractBuildGradle<*, *>, dependencies: List<Dependency>) :
    BuildSystem(buildGradle, dependencies) {

    companion object {
        /**
         * 根据文件名构建gradle build system
         */
        @JvmStatic
        @SneakyThrows
        fun from(context: DataContext): GradleBuildSystem {
            val psiFile = context.getData(CommonDataKeys.PSI_FILE)
            val project = context.getData(CommonDataKeys.PROJECT)
            if (psiFile == null || project == null) {
                throw ShowErrorException.internal()
            }

            val name = psiFile.name
            val buildGradle: AbstractBuildGradle<*, *> =
                when (name) {
                    GradleConstants.DEFAULT_SCRIPT_NAME -> BuildGradle(project, (psiFile as GroovyFile?)!!)
                    GradleConstants.KOTLIN_DSL_SCRIPT_NAME -> BuildGradleKts(project, (psiFile as KtFile?)!!)
                    else -> throw ShowErrorException("Not support extension!")
                }
            if (StringUtils.isEmpty(project.basePath)) {
                throw ShowErrorException.internal()
            }

            val projectData =
                ExternalSystemApiUtil.findProjectNode(project, GradleConstants.SYSTEM_ID, project.basePath!!)
                    ?: throw ShowErrorException.internal()
            val dependencies = projectData.children.stream()
                .filter { node: DataNode<*> -> ProjectKeys.LIBRARY == node.key }
                .map { node: DataNode<*> -> node.data as LibraryData }
                .map { lib: LibraryData -> Dependency(lib.groupId, lib.artifactId, lib.version) }
                .collect(Collectors.toList())

            return GradleBuildSystem(buildGradle, dependencies)
        }
    }
}